### ISSUE TITLE
[Bug]: Consistency of platform-specific methods: no-op vs error

### ISSUE THREAD

**Issue by jtbandes** [electron/electron#28662]
### Preflight Checklist

I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Electron Version

12.0.0

### What operating system are you using?

macOS

### Operating System Version

11.2.3

### What arch are you using?

arm64 (including Apple Silicon)

### Expected Behavior

Methods that are platform-specific, such as `app.dock` and `systemPreferences.getUserDefault`, should either be complete no-ops on other platforms, or should be missing/undefined, or trying to call them should throw an error.

If an error is to be thrown, ideally rather than "cannot read property 'x' of undefined", the user would get an explicit error such as "this method is only available on [platform]".

### Actual Behavior

As discussed [in this issue](https://github.com/electron/typescript-definitions/pull/183#pullrequestreview-635161043), currently there is some inconsistency between various platform-specific methods. The TypeScript definitions advertise that these methods are **not** optional, leading to runtime crashes that otherwise the TS compiler could've helped avoid. If the decision is to make these methods missing/errors on other platforms, then the TS definitions can be updated to help users avoid calling them on the wrong platform (actually this could be done even if they are no-ops).

**Comment by nornagon** [comment]
I agree it would be nice for these to be consistent. As to whether platform-specific methods should no-op, throw or be absent, I think that bears some discussion. Let's leave aside here _options_ that are platform-specific (such as `extraSize` in [`BrowserWindow.setAspectRatio`](https://www.electronjs.org/docs/api/browser-window#winsetaspectratioaspectratio-extrasize) or `visibleOnFullscreen` in [`BrowserWindow.setVisibleOnAllWorkspaces`](https://www.electronjs.org/docs/api/browser-window#winsetvisibleonallworkspacesvisible-options)) and restrict discussion to methods and properties which are platform-specific, such as [`BrowserWindow.setContentProtection`](https://www.electronjs.org/docs/api/browser-window#winsetcontentprotectionenable-macos-windows) and [`BrowserWindow.isDocumentEdited`](https://www.electronjs.org/docs/api/browser-window#winisdocumentedited-macos).

There are three main possibilities, I think. on unsupported platforms, the method or property could:
1. be `undefined`,
2. throw an exception if used, or
3. feign success or otherwise be undisruptive (no-op)

Each of these has merits and drawbacks.

#### 1. `undefined`

- `+` it's possible to test for presence at runtime without having to cross-reference the docs to check what platforms are supported. e.g. `thing.platformSpecificFunction?()` or `if (thing.platformSpecificFunction) ...`.
- `+` fails noisily; it will be clear that the feature isn't present on the platform if the app attempts to use it.
- `+` if using TypeScript, your IDE will warn you about the non-universal nature of the method.
- `-` crappy error message, `undefined is not a function` doesn't say anything about OS specificity.
- `-` disruptive when the feature is incidental to the functioning of the app.

#### 2. throw an exception

- `+` fails noisily; it will be clear that the feature isn't present on the platform if the app attempts to use it.
- `+` it's immediately clear from the exception what the problem is.
- `-` harder to feature-test automatically—instead of `if (thing.platformSpecificFunction)` you need `if (process.platform === 'darwin' || process.platform === 'linux') ...`, and you'll need to update that if a future electron version adds/removes support for a particular platform.
- `-` disruptive when the feature is incidental to the functioning of the app.

#### 3. no-op

- `+` for features that are incidental to the functioning of the app (i.e. are not required for the app to do its job), having the unsupported operation fail silently means the rest of the app will continue to function, instead of crashing.
- `+` there's nothing for the app developer to test or change to support other platforms.
- `-` harder to feature-test automatically—instead of `if (thing.platformSpecificFunction)` you need `if (process.platform === 'darwin' || process.platform === 'linux') ...`, and you'll need to update that if a future electron version adds/removes support for a particular platform.
- `-` for APIs which return a value, it's unclear what the "no-op" version should return.

**Comment by zcbenz** [comment]
Whichever way we take, should we change existing APIs to be consistent? It will be a breaking change and apps will have a difficult time finding and fixing code that rely on old assumptions.

**Comment by nornagon** [comment]
Hm, yeah, I think it's probably not worth making any breaking changes for the sake of consistency. Not all of the possible changes are breaking, though.

#### breaking types
* changing from undefined -> exception or no-op breaks feature-testing by `if (thing.method)`
* changing from no-op -> throw or undefined breaks apps that call that method on unsupported platforms

#### non-breaking types
* changing from throw -> undefined is non-breaking
* changing from throw -> no-op is non-breaking

unfortunately this graph is not well-ordered, so there's no top state we can settle to without potentially breaking some apps.

So the outcome of this should be a guideline change for future APIs, not necessarily changes to existing APIs. We should probably also clearly document the behavior of existing APIs when called on unsupported platforms.

**Comment by jtbandes** [comment]
Perhaps it would be worth making "breaking" changes to the TypeScript definitions (changes that will cause the TS compiler to reject some existing code) even if the underlying method implementations don't change today? That is, methods which would throw when called can be removed or marked optional in the TS defs. See more discussion of possible approaches to platform guards in https://github.com/electron/typescript-definitions/pull/183#issuecomment-819213137
